From d36f64e6a7cfc3d4970ea0c271045efd9bc2f403 Mon Sep 17 00:00:00 2001
From: Camelia Groza <camelia.groza@nxp.com>
Date: Mon, 15 Jan 2018 17:41:01 +0200
Subject: [PATCH 183/741] sdk_dpaa: update buffer recycling conditions

commit 83f1387b42e0dd290634eea615ba8d5822649c19 from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Guarantee there is enough space inside the skb's headroom to store the
skb back-pointer before recycling the buffer. The back-pointer is stored
right before the buffer's start.

Signed-off-by: Camelia Groza <camelia.groza@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_sg.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_sg.c b/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_sg.c
index 32e62e6d91da..f73743986183 100644
--- a/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_sg.c
+++ b/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_sg.c
@@ -350,7 +350,13 @@ bool dpa_buf_is_recyclable(struct sk_buff *skb,
 	/* left align to the nearest cacheline */
 	new = (unsigned char *)((unsigned long)new & ~(SMP_CACHE_BYTES - 1));
 
-	if (likely(new >= skb->head &&
+	/* Make sure there is enough space to store the skb back-pointer in
+	 * the headroom, right before the start of the buffer.
+	 *
+	 * Guarantee that both maximum size and maximum data offsets aren't
+	 * crossed.
+	 */
+	if (likely(new >= (skb->head + sizeof(void *)) &&
 		   new >= (skb->data - DPA_MAX_FD_OFFSET) &&
 		   skb_end_pointer(skb) - new <= DPA_RECYCLE_MAX_SIZE)) {
 		*new_buf_start = new;
-- 
2.17.1

